### PR TITLE
GoogleVertexAIImageEmbeddings

### PR DESCRIPTION
**Pull request by afirstenberg**
This provides an implementation of a class that works with Google's Vertex AI ImageEmbedding model.
It can take text, a buffer with an image, or both, and create embeddings for them in the same embedding space.

See https://console.cloud.google.com/vertex-ai/publishers/google/model-garden/5 for the Model Garden card about this model and https://cloud.google.com/blog/products/ai-machine-learning/enterprise-ready-generative-ai-models-go-ga-in-vertex-ai for the announcement.

@jacoblee93

### CODE REVIEW

File: langchain/src/embeddings/googlevertexai-image.ts [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+      image: r.imageEmbedding,
+    }));
+  }
+
+  async embedMedia(media: GoogleVertexAIMedia[]): Promise<MediaEmbeddings[]> {
+    // Only one media embedding request is allowed
+    return Promise.all(media.map((m) => this.embedMediaQuery(m)));
+  }
+
+  async embedMediaQuery(media: GoogleVertexAIMedia): Promise<MediaEmbeddings> {
```
Comment (jacoblee93) [inline langchain/src/embeddings/googlevertexai-image.ts:103]: Is there a purpose to embedding both at the same time? Does one input influence the result of the other?

**Review by jacoblee93 — COMMENTED** [review]
Thanks for submitting this - I think this is generally fine as a one-off but would like to come up with a strategy for how this looks downstream on vector store retrieval as well as document transformation (e.g. text splitting obviously doesn't make sense on images). 

Please give me a bit more time.

File: langchain/src/embeddings/googlevertexai-image.ts (hunk 1 above)
Comment (afirstenberg) [reply langchain/src/embeddings/googlevertexai-image.ts:103]: The API allows for it is the biggest reason. The two vectors are independent of each other, except they live in the same vector space.

My guess is that the original purpose was so you could submit an image and a text, get the resultant vectors, and compute the similarity between the two all in one shot - which should be faster than two queries. I saw at least one other model that did this as well.

**Comment by afirstenberg** [comment]
I would agree - I'd love to see a more general purpose Document that could hold either text or images or...

...but I think that will take an extensive restructuring of the code across a wide range of classes.
It's a conversation we should stort.
But I don't think we should hold this class till then.

File: langchain/scripts/create-entrypoints.js [hunk 2]
```diff
@@ -40,6 +40,7 @@ const entrypoints = {
   "embeddings/tensorflow": "embeddings/tensorflow",
   "embeddings/hf": "embeddings/hf",
   "embeddings/googlevertexai": "embeddings/googlevertexai",
+  "embeddings/googlevertexai-image": "embeddings/googlevertexai-image",
```
Comment (jacoblee93) [inline langchain/scripts/create-entrypoints.js:43]: Underscore is convention vs. hyphen

File: langchain/src/embeddings/googlevertexai-image.ts [hunk 3]
```diff
@@ -0,0 +1,143 @@
+import { Embeddings, EmbeddingsParams } from "./base.js";
+import {
+  GoogleVertexAIBasePrediction,
+  GoogleVertexAIConnectionParams,
+  GoogleVertexAILLMResponse,
+} from "../types/googlevertexai-types.js";
+import { GoogleVertexAIConnection } from "../util/googlevertexai-connection.js";
+import { AsyncCallerCallOptions } from "../util/async_caller.js";
+// import { chunkArray } from "../util/chunk.js";
```
Comment (jacoblee93) [inline langchain/src/embeddings/googlevertexai-image.ts:9]: Remove comment

File: langchain/src/embeddings/googlevertexai-image.ts [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+      text?: string;
+      image: Buffer;
+    };
+
+export type MediaEmbeddings = {
+  text?: number[];
+  image?: number[];
+};
+
+export class GoogleVertexAIImageEmbeddings
```
Comment (jacoblee93) [inline langchain/src/embeddings/googlevertexai-image.ts:48]: Do we think it's premature to call this `GoogleVertexAIMultimodalEmbeddings`?

File: langchain/src/embeddings/googlevertexai-image.ts (hunk 1 above)
Comment (jacoblee93) [reply langchain/src/embeddings/googlevertexai-image.ts:103]: I think I'd rather keep them as separate methods and add the combined one in if there's some benefit to reduce confusion for now - this one is already going to be an odd duck.

**Review by jacoblee93 — COMMENTED** [review]
A few comments!

File: langchain/src/embeddings/googlevertexai-image.ts (hunk 4 above)
Comment (afirstenberg) [reply langchain/src/embeddings/googlevertexai-image.ts:48]: I think when I started developing it, they were ambiguous about what its name would be. But they seem to have (mostly) settled on using "multimodal". I'll change the name.

File: langchain/scripts/create-entrypoints.js (hunk 2 above)
Comment (afirstenberg) [reply langchain/scripts/create-entrypoints.js:43]: Will change, tho this was based on llms/openai-chat

**Comment by afirstenberg** [comment]
> A few comments!

These issues should be addressed now.

**Comment by jacoblee93** [comment]
Hey @afirstenberg, I've moved this to `experimental/multimodal_embeddings/googlevertexai` for now so that folks know that breaking changes may occur here in the future as things develop. Otherwise looks great, thank you so much!

**Comment by jacoblee93** [comment]
Added a more end-to-end example with a vector store. Didn't seem to return the image when I searched for `"bird"` but image query -> image seemed to work reasonably well.

**Pull request merged**